File: KargoPaylasimJava/app/src/main/java/com/example/kargopaylasimjava/viewmodel/ShipmentViewModel.java
package com.example.kargopaylasimjava.viewmodel;

import androidx.annotation.NonNull;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.example.kargopaylasimjava.dto.ApiResp;
import com.example.kargopaylasimjava.dto.ReceiverDtos;
import com.example.kargopaylasimjava.dto.ShipmentDtos;
import com.example.kargopaylasimjava.model.UiState;
import com.example.kargopaylasimjava.repo.CargoRepository;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class ShipmentViewModel extends ViewModel {

    private final CargoRepository repo;

    public MutableLiveData<UiState<List<ShipmentDtos.ShipmentDto>>> listState = new MutableLiveData<>(UiState.idle());
    public MutableLiveData<UiState<ShipmentDtos.ShipmentCreateResp>> createState = new MutableLiveData<>(UiState.idle());
    public MutableLiveData<UiState<ShipmentDtos.ShipmentDetailDto>> detailState = new MutableLiveData<>(UiState.idle());
    public MutableLiveData<UiState<ReceiverDtos.ReceiverLookupResp>> lookupState = new MutableLiveData<>(UiState.idle());
    public MutableLiveData<UiState<ShipmentDtos.ShipmentRegenerateResp>> regenerateState = new MutableLiveData<>(UiState.idle());
    public MutableLiveData<UiState<ShipmentDtos.ShipmentDeleteResp>> deleteState = new MutableLiveData<>(UiState.idle());

    public ShipmentViewModel(CargoRepository repo) {
        this.repo = repo;
    }

    public void loadShipments() {
        listState.setValue(UiState.loading());

        repo.shipmentList("all").enqueue(new Callback<ApiResp<ShipmentDtos.ShipmentListResp>>() {
            @Override
            public void onResponse(@NonNull Call<ApiResp<ShipmentDtos.ShipmentListResp>> call,
                                   @NonNull Response<ApiResp<ShipmentDtos.ShipmentListResp>> response) {
                ApiResp<ShipmentDtos.ShipmentListResp> body = response.body();
                if (response.isSuccessful() && body != null && body.ok && body.data != null) {
                    listState.setValue(UiState.success(body.data.shipments));
                } else {
                    String msg = (body != null && body.error != null) ? body.error : "Load failed";
                    listState.setValue(UiState.error(msg));
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiResp<ShipmentDtos.ShipmentListResp>> call, @NonNull Throwable t) {
                listState.setValue(UiState.error(t.getMessage() != null ? t.getMessage() : "Network error"));
            }
        });
    }

    public void lookupReceiver(String phone) {
        lookupState.setValue(UiState.loading());

        repo.receiverLookup(phone).enqueue(new Callback<ApiResp<ReceiverDtos.ReceiverLookupResp>>() {
            @Override
            public void onResponse(@NonNull Call<ApiResp<ReceiverDtos.ReceiverLookupResp>> call,
                                   @NonNull Response<ApiResp<ReceiverDtos.ReceiverLookupResp>> response) {
                ApiResp<ReceiverDtos.ReceiverLookupResp> body = response.body();
                if (response.isSuccessful() && body != null && body.ok && body.data != null) {
                    lookupState.setValue(UiState.success(body.data));
                } else {
                    String msg = (body != null && body.error != null) ? body.error : "Lookup failed";
                    lookupState.setValue(UiState.error(msg));
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiResp<ReceiverDtos.ReceiverLookupResp>> call, @NonNull Throwable t) {
                lookupState.setValue(UiState.error(t.getMessage() != null ? t.getMessage() : "Network error"));
            }
        });
    }

    public void createShipment(String receiverPhone, Integer senderAddressId) {
        createState.setValue(UiState.loading());

        repo.shipmentCreate(receiverPhone, senderAddressId).enqueue(new Callback<ApiResp<ShipmentDtos.ShipmentCreateResp>>() {
            @Override
            public void onResponse(@NonNull Call<ApiResp<ShipmentDtos.ShipmentCreateResp>> call,
                                   @NonNull Response<ApiResp<ShipmentDtos.ShipmentCreateResp>> response) {
                ApiResp<ShipmentDtos.ShipmentCreateResp> body = response.body();
                if (response.isSuccessful() && body != null && body.ok && body.data != null) {
                    createState.setValue(UiState.success(body.data));
                } else {
                    String msg = (body != null && body.error != null) ? body.error : "Create failed";
                    createState.setValue(UiState.error(msg));
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiResp<ShipmentDtos.ShipmentCreateResp>> call, @NonNull Throwable t) {
                createState.setValue(UiState.error(t.getMessage() != null ? t.getMessage() : "Network error"));
            }
        });
    }

    public void loadDetail(int id) {
        detailState.setValue(UiState.loading());

        repo.shipmentDetail(id).enqueue(new Callback<ApiResp<ShipmentDtos.ShipmentDetailResp>>() {
            @Override
            public void onResponse(@NonNull Call<ApiResp<ShipmentDtos.ShipmentDetailResp>> call,
                                   @NonNull Response<ApiResp<ShipmentDtos.ShipmentDetailResp>> response) {
                ApiResp<ShipmentDtos.ShipmentDetailResp> body = response.body();
                if (response.isSuccessful() && body != null && body.ok && body.data != null && body.data.shipment != null) {
                    detailState.setValue(UiState.success(body.data.shipment));
                } else {
                    String msg = (body != null && body.error != null) ? body.error : "Detail failed";
                    detailState.setValue(UiState.error(msg));
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiResp<ShipmentDtos.ShipmentDetailResp>> call, @NonNull Throwable t) {
                detailState.setValue(UiState.error(t.getMessage() != null ? t.getMessage() : "Network error"));
            }
        });
    }

    public void regenerateCode(int shipmentId) {
        regenerateState.setValue(UiState.loading());

        repo.shipmentRegenerate(shipmentId).enqueue(new Callback<ApiResp<ShipmentDtos.ShipmentRegenerateResp>>() {
            @Override
            public void onResponse(@NonNull Call<ApiResp<ShipmentDtos.ShipmentRegenerateResp>> call,
                                   @NonNull Response<ApiResp<ShipmentDtos.ShipmentRegenerateResp>> response) {
                ApiResp<ShipmentDtos.ShipmentRegenerateResp> body = response.body();
                if (response.isSuccessful() && body != null && body.ok && body.data != null) {
                    regenerateState.setValue(UiState.success(body.data));
                } else {
                    String msg = (body != null && body.error != null) ? body.error : "Regenerate failed";
                    regenerateState.setValue(UiState.error(msg));
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiResp<ShipmentDtos.ShipmentRegenerateResp>> call, @NonNull Throwable t) {
                regenerateState.setValue(UiState.error(t.getMessage() != null ? t.getMessage() : "Network error"));
            }
        });
    }

    public void deleteShipment(int shipmentId) {
        deleteState.setValue(UiState.loading());

        repo.shipmentDelete(shipmentId).enqueue(new Callback<ApiResp<ShipmentDtos.ShipmentDeleteResp>>() {
            @Override
            public void onResponse(@NonNull Call<ApiResp<ShipmentDtos.ShipmentDeleteResp>> call,
                                   @NonNull Response<ApiResp<ShipmentDtos.ShipmentDeleteResp>> response) {
                ApiResp<ShipmentDtos.ShipmentDeleteResp> body = response.body();
                if (response.isSuccessful() && body != null && body.ok && body.data != null) {
                    deleteState.setValue(UiState.success(body.data));
                } else {
                    String msg = (body != null && body.error != null) ? body.error : "Delete failed";
                    deleteState.setValue(UiState.error(msg));
                }
            }

            @Override
            public void onFailure(@NonNull Call<ApiResp<ShipmentDtos.ShipmentDeleteResp>> call, @NonNull Throwable t) {
                deleteState.setValue(UiState.error(t.getMessage() != null ? t.getMessage() : "Network error"));
            }
        });
    }
}

File: KargoPaylasimKotlin/app/src/main/java/com/example/kargopaylasimkotlin/viewmodel/ShipmentViewModel.kt
package com.example.kargopaylasimkotlin.viewmodel


import androidx.lifecycle.*
import com.example.kargopaylasimkotlin.dto.ReceiverLookupResp
import com.example.kargopaylasimkotlin.dto.ShipmentCreateResp
import com.example.kargopaylasimkotlin.dto.ShipmentDeleteResp
import com.example.kargopaylasimkotlin.dto.ShipmentDetailDto
import com.example.kargopaylasimkotlin.dto.ShipmentDto
import com.example.kargopaylasimkotlin.dto.ShipmentItemDto
import com.example.kargopaylasimkotlin.dto.ShipmentRegenerateResp
import com.example.kargopaylasimkotlin.model.UiState
import com.example.kargopaylasimkotlin.repo.CargoRepository

import kotlinx.coroutines.launch

class ShipmentViewModel(private val repo: CargoRepository) : ViewModel() {

    val listState = MutableLiveData<UiState<List<ShipmentDto>>>(UiState.Idle)
    val createState = MutableLiveData<UiState<ShipmentCreateResp>>(UiState.Idle)
    val detailState = MutableLiveData<UiState<ShipmentDetailDto>>(UiState.Idle)
    val lookupState = MutableLiveData<UiState<ReceiverLookupResp>>(UiState.Idle)
    val regenerateState = MutableLiveData<UiState<ShipmentRegenerateResp>>(UiState.Idle)
    val deleteState = MutableLiveData<UiState<ShipmentDeleteResp>>(UiState.Idle)

    fun deleteShipment(shipmentId: Int) {
        deleteState.value = UiState.Loading
        viewModelScope.launch {
            try {
                val resp = repo.shipmentDelete(shipmentId)
                if (resp.ok && resp.data != null) deleteState.value = UiState.Success(resp.data)
                else deleteState.value = UiState.Error(resp.error ?: "Delete failed")
            } catch (e: Exception) {
                deleteState.value = UiState.Error(e.message ?: "Network error")
            }
        }
    }

    fun loadShipments() {
        listState.value = UiState.Loading
        viewModelScope.launch {
            try {

                val resp = repo.shipmentList(type = "all")
                android.util.Log.d("SHIP", "count=" + (resp.data?.shipments?.size ?: -1))
                if (resp.ok && resp.data != null) {
                    listState.value = UiState.Success(resp.data.shipments) // ✅ burası
                } else {
                    listState.value = UiState.Error(resp.error ?: "Load failed")
                }
            } catch (e: Exception) {
                listState.value = UiState.Error(e.message ?: "Network error")
            }
        }
    }


    fun lookupReceiver(phone: String) {
        lookupState.value = UiState.Loading
        viewModelScope.launch {
            try {
                val resp = repo.receiverLookup(phone)
                if (resp.ok && resp.data != null) lookupState.value = UiState.Success(resp.data)
                else lookupState.value = UiState.Error(resp.error ?: "Lookup failed")
            } catch (e: Exception) {
                lookupState.value = UiState.Error(e.message ?: "Network error")
            }
        }
    }

    fun createShipment(receiverPhone: String, senderAddressId: Int?) {
        createState.value = UiState.Loading
        viewModelScope.launch {
            try {
                val resp = repo.shipmentCreate(receiverPhone, senderAddressId)
                if (resp.ok && resp.data != null) createState.value = UiState.Success(resp.data)
                else createState.value = UiState.Error(resp.error ?: "Create failed")
            } catch (e: Exception) {
                createState.value = UiState.Error(e.message ?: "Network error")
            }
        }
    }

    fun loadDetail(id: Int) {
        detailState.value = UiState.Loading
        viewModelScope.launch {
            try {
                val resp = repo.shipmentDetail(id)
                if (resp.ok && resp.data != null) detailState.value = UiState.Success(resp.data.shipment)
                else detailState.value = UiState.Error(resp.error ?: "Detail failed")
            } catch (e: Exception) {
                detailState.value = UiState.Error(e.message ?: "Network error")
            }
        }
    }

    fun regenerateCode(shipmentId: Int) {
        regenerateState.value = UiState.Loading
        viewModelScope.launch {
            try {
                val resp = repo.shipmentRegenerate(shipmentId)
                if (resp.ok && resp.data != null) regenerateState.value = UiState.Success(resp.data)
                else regenerateState.value = UiState.Error(resp.error ?: "Regenerate failed")
            } catch (e: Exception) {
                regenerateState.value = UiState.Error(e.message ?: "Network error")
            }
        }
    }
}
